### PR TITLE
Add field obfuscation feature

### PR DESCRIPTION
**Pull request by Ismaley**
Hello @kishorebanala, @apanicker-nflx, @cyzhao 

As it was discussed in this issue #976, we've been needing to hide sensitive information from the workflows executed. Back then we needed only to hide information from logs but now we came to a point where we also needed to hide certain fields from the workflow and from the tasks executed and we came up with a solution.

A brief explanation of what I've done: 
- a point-to-point system using dynoQueues to process the field obfuscation, so it won't add extra time to the workflow's execution and in case of system failures it'll also prevent data inconsistency.

- a new workflow resource where you can start the obfuscation process for all workflows with a given name and version. This can help fix inconsistencies where the user fails to provide the exact path of the json to be replaced (the workflowDef must be updated before calling this resource).

I'd be glad if you could review this pull request and also if you could add some thoughts on what could be improved here. 

Thanks in advance.

### CODE REVIEW

File: core/src/test/resources/test_workflow.json [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+        "startDelay": 0,
+        "optional": false
+      }
+    ],
+    "schemaVersion": 2,
+    "restartable": true,
+    "workflowStatusListenerEnabled": false,
+    "obfuscationFields": {
+      "workflowFields": "output.response.body",
+      "taskFields": "Request02.inputData.http_request.headers.x-workflow-id"
```
Comment (kishorebanala) [inline core/src/test/resources/test_workflow.json]: I see the fields are comma separated. And the implementation doesn't have a trim, to avoid accidental spaces after comma. It might be best to have json arrays of tokens. Eg: 

```json
"obfuscationFields": {
      "workflowFields": [
            "output.response.body",
            "output.someKey"
      ],
      "taskFields": [
            "Request02.inputData.http_request.headers.x-workflow-id",
            "Request04.outputData.someValue"
      ]
 }
```

**Comment by Ismaley** [comment]
@kishorebanala and @apanicker-nflx, thanks for reviewing this PR.

@kishorebanala, I liked your idea! Just to make sure, are you suggesting something like it was done with the elasticsearch modules, whether you can opt which one you're going to install based on the configuration or to port the implementations to the contribs module and leave just the interfaces on the core and server modules?

I'll add the documentation once I get the changes done.

File: core/src/test/resources/test_workflow.json (hunk 1 above)
Comment (Ismaley) [reply core/src/test/resources/test_workflow.json]: Actually at the first attempt to model this new field I opted for a Map<String, <List<String>> but I got some errors with gradle's protogen task, it seems the lib used has some struggles with nested types. When I switched to a Map<String, String> it worked normally.
Maybe I missed some additional configuration or if you know there's a workaround for this I'd be glad to change the field type, otherwise I'll add a trim to the implementation.

**Comment by apanicker-nflx** [comment]
> @kishorebanala and @apanicker-nflx, thanks for reviewing this PR.
> 
> @kishorebanala, I liked your idea! Just to make sure, are you suggesting something like it was done with the elasticsearch modules, whether you can opt which one you're going to install based on the configuration or to port the implementations to the contribs module and leave just the interfaces on the core and server modules?
> 
> I'll add the documentation once I get the changes done.

Please create a new module akin to the elasticsearch modules, so that users can choose to install this module if required.

**Comment by Ismaley** [comment]
Greetings, I just wanted to notify that I'm currently working on this PR changes along with some improvements and other changes I noticed that will be needed. I'll push the changes as soon as I finish them. 
Thanks for the reviews and ideas so far :)

File: core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java [hunk 2]
```diff
@@ -44,32 +41,18 @@
...
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.Collections;
-import java.util.Comparator;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
-import java.util.Objects;
-import java.util.Optional;
+import java.util.*;
```
Comment (apanicker-nflx) [inline core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java]: Please avoid using wildcard imports.

File: core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java [hunk 3]
```diff
@@ -18,11 +18,8 @@
 import com.netflix.conductor.common.metadata.tasks.PollData;
 import com.netflix.conductor.common.metadata.tasks.Task;
 import com.netflix.conductor.common.metadata.tasks.TaskResult;
-import com.netflix.conductor.common.metadata.workflow.RerunWorkflowRequest;
-import com.netflix.conductor.common.metadata.workflow.SkipTaskRequest;
-import com.netflix.conductor.common.metadata.workflow.TaskType;
-import com.netflix.conductor.common.metadata.workflow.WorkflowDef;
-import com.netflix.conductor.common.metadata.workflow.WorkflowTask;
+import com.netflix.conductor.common.metadata.workflow.*;
```
Comment (apanicker-nflx) [inline core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java]: Please avoid using wildcard imports.

File: core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java [hunk 4]
```diff
@@ -44,32 +41,18 @@
...
+
+import static com.netflix.conductor.common.metadata.tasks.Task.Status.*;
 import static com.netflix.conductor.common.metadata.tasks.Task.Status.valueOf;
 import static com.netflix.conductor.common.metadata.workflow.TaskType.SUB_WORKFLOW;
 import static com.netflix.conductor.common.utils.ExternalPayloadStorage.PayloadType.TASK_OUTPUT;
 import static com.netflix.conductor.common.utils.ExternalPayloadStorage.PayloadType.WORKFLOW_INPUT;
-import static com.netflix.conductor.core.execution.ApplicationException.Code.CONFLICT;
-import static com.netflix.conductor.core.execution.ApplicationException.Code.INVALID_INPUT;
-import static com.netflix.conductor.core.execution.ApplicationException.Code.NOT_FOUND;
+import static com.netflix.conductor.core.execution.ApplicationException.Code.*;
```
Comment (apanicker-nflx) [inline core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java]: Please avoid using wildcard imports.

File: core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java [hunk 5]
```diff
@@ -1054,6 +1048,28 @@ public void executeSystemTask(WorkflowSystemTask systemTask, String taskId, int
...
+    private WorkflowDef getWorkflowDef(String name, Integer version) {
+        Optional<WorkflowDef> workflowDef = metadataDAO.get(name, version);
+        if(workflowDef.isPresent()) {
+            return workflowDef.get();
+        } else {
+            throw new ApplicationException(Code.INVALID_INPUT, "workflowDefinition does not exists");
+        }
+    }
+
+    private SearchResult<String> searchWorkflowsToPublish(String name, Integer version, int start) {
```
Comment (apanicker-nflx) [inline core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java]: Is this method being used?

File: core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutorModule.java [hunk 6]
```diff
@@ -23,12 +12,16 @@
     protected void configure() {
         bind(WorkflowStatusListener.class).to(WorkflowStatusListenerStub.class);//default implementation
 
+//        bind(WorkflowObfuscationQueuePublisher.class).asEagerSingleton();
```
Comment (apanicker-nflx) [inline core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutorModule.java]: Please remove commented out code.

File: core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutorModule.java [hunk 7]
```diff
@@ -1,18 +1,7 @@
...
-import com.netflix.conductor.service.EventServiceImpl;
-import com.netflix.conductor.service.MetadataService;
-import com.netflix.conductor.service.MetadataServiceImpl;
-import com.netflix.conductor.service.TaskService;
-import com.netflix.conductor.service.TaskServiceImpl;
-import com.netflix.conductor.service.WorkflowBulkService;
-import com.netflix.conductor.service.WorkflowBulkServiceImpl;
-import com.netflix.conductor.service.WorkflowService;
-import com.netflix.conductor.service.WorkflowServiceImpl;
+import com.netflix.conductor.service.*;
```
Comment (apanicker-nflx) [inline core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutorModule.java]: Please avoid using wildcard imports.

File: core/src/main/java/com/netflix/conductor/core/execution/WorkflowObfuscationQueuePublisher.java [hunk 8]
```diff
@@ -0,0 +1,14 @@
+package com.netflix.conductor.core.execution;
+
+import com.netflix.conductor.common.metadata.workflow.WorkflowDef;
+import com.netflix.conductor.common.run.Workflow;
+
+import java.util.List;
+
+public interface WorkflowObfuscationQueuePublisher {
```
Comment (apanicker-nflx) [inline core/src/main/java/com/netflix/conductor/core/execution/WorkflowObfuscationQueuePublisher.java]: Please add javadocs at both interface level and method level.

File: core/src/main/java/com/netflix/conductor/core/execution/WorkflowObfuscationQueuePublisherStub.java [hunk 9]
```diff
@@ -0,0 +1,25 @@
+package com.netflix.conductor.core.execution;
+
+import com.netflix.conductor.common.metadata.workflow.WorkflowDef;
+import com.netflix.conductor.common.run.Workflow;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+
+public class WorkflowObfuscationQueuePublisherStub implements WorkflowObfuscationQueuePublisher {
```
Comment (apanicker-nflx) [inline core/src/main/java/com/netflix/conductor/core/execution/WorkflowObfuscationQueuePublisherStub.java]: Please add javadoc.

File: core/src/test/java/com/netflix/conductor/core/execution/TestWorkflowExecutor.java [hunk 10]
```diff
@@ -25,19 +25,7 @@
...
-import com.netflix.conductor.core.execution.mapper.ForkJoinTaskMapper;
-import com.netflix.conductor.core.execution.mapper.HTTPTaskMapper;
-import com.netflix.conductor.core.execution.mapper.JoinTaskMapper;
-import com.netflix.conductor.core.execution.mapper.LambdaTaskMapper;
-import com.netflix.conductor.core.execution.mapper.SimpleTaskMapper;
-import com.netflix.conductor.core.execution.mapper.SubWorkflowTaskMapper;
-import com.netflix.conductor.core.execution.mapper.TaskMapper;
-import com.netflix.conductor.core.execution.mapper.UserDefinedTaskMapper;
-import com.netflix.conductor.core.execution.mapper.WaitTaskMapper;
+import com.netflix.conductor.core.execution.mapper.*;
```
Comment (apanicker-nflx) [inline core/src/test/java/com/netflix/conductor/core/execution/TestWorkflowExecutor.java]: Please avoid using wildcard imports.

File: core/src/test/java/com/netflix/conductor/core/execution/TestWorkflowExecutor.java [hunk 11]
```diff
@@ -50,41 +38,24 @@
...
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.Collections;
-import java.util.HashMap;
-import java.util.LinkedList;
-import java.util.List;
-import java.util.Map;
-import java.util.Optional;
-import java.util.UUID;
+import java.util.*;
```
Comment (apanicker-nflx) [inline core/src/test/java/com/netflix/conductor/core/execution/TestWorkflowExecutor.java]: Please avoid using wildcard imports.

File: core/src/test/java/com/netflix/conductor/core/execution/TestWorkflowExecutor.java [hunk 12]
```diff
@@ -50,41 +38,24 @@
...
+import static java.util.Collections.singletonList;
 import static java.util.Comparator.comparingInt;
 import static java.util.stream.Collectors.groupingBy;
 import static java.util.stream.Collectors.maxBy;
-import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.assertFalse;
-import static org.junit.Assert.assertNotNull;
-import static org.junit.Assert.assertNull;
-import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.*;
```
Comment (apanicker-nflx) [inline core/src/test/java/com/netflix/conductor/core/execution/TestWorkflowExecutor.java]: Please avoid using wildcard imports.

File: core/src/test/java/com/netflix/conductor/core/execution/TestWorkflowExecutor.java [hunk 13]
```diff
@@ -50,41 +38,24 @@
...
 import static org.mockito.Matchers.anyString;
 import static org.mockito.Matchers.eq;
-import static org.mockito.Mockito.doAnswer;
-import static org.mockito.Mockito.doNothing;
-import static org.mockito.Mockito.mock;
-import static org.mockito.Mockito.never;
-import static org.mockito.Mockito.times;
-import static org.mockito.Mockito.verify;
-import static org.mockito.Mockito.when;
+import static org.mockito.Mockito.*;
```
Comment (apanicker-nflx) [inline core/src/test/java/com/netflix/conductor/core/execution/TestWorkflowExecutor.java]: Please avoid using wildcard imports.

File: conductor-obfuscation/build.gradle [hunk 14]
```diff
@@ -0,0 +1,15 @@
+plugins {
```
Comment (apanicker-nflx) [inline conductor-obfuscation/build.gradle]: Is this file needed?

File: obfuscation/src/main/java/com/netflix/conductor/obfuscation/service/ObfuscationResourceService.java [hunk 15]
```diff
@@ -0,0 +1,7 @@
+package com.netflix.conductor.obfuscation.service;
+
+public interface ObfuscationResourceService {
```
Comment (apanicker-nflx) [inline obfuscation/src/main/java/com/netflix/conductor/obfuscation/service/ObfuscationResourceService.java]: Please add javadoc

File: core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java [hunk 16]
```diff
@@ -1054,6 +1048,28 @@ public void executeSystemTask(WorkflowSystemTask systemTask, String taskId, int
         }
     }
 
+    private void validateInput(String name, Integer version) {
```
Comment (apanicker-nflx) [inline core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java]: Is this method being used?

File: core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java [hunk 17]
```diff
@@ -1054,6 +1048,28 @@ public void executeSystemTask(WorkflowSystemTask systemTask, String taskId, int
...
+        if(version == null || version <= 0) {
+            throw new ApplicationException(Code.INVALID_INPUT, "workflow version must be present and cannot have 0 or below value");
+        }
+        if(isEmpty(name)) {
+            throw new ApplicationException(Code.INVALID_INPUT, "workflow name must be present");
+        }
+    }
+
+    private WorkflowDef getWorkflowDef(String name, Integer version) {
```
Comment (apanicker-nflx) [inline core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java]: Is this method being used?

File: obfuscation/src/main/java/com/netflix/conductor/obfuscation/service/ObfuscationServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,185 @@
...
+import com.netflix.conductor.obfuscation.exception.ObfuscationServiceException;
+import com.netflix.conductor.service.MetadataService;
+import net.minidev.json.JSONArray;
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.util.*;
```
Comment (apanicker-nflx) [inline obfuscation/src/main/java/com/netflix/conductor/obfuscation/service/ObfuscationServiceImpl.java]: Please avoid using wildcard imports.

File: obfuscation/src/main/java/com/netflix/conductor/obfuscation/publisher/WorkflowObfuscationQueuePublisher.java [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+import com.netflix.conductor.dao.QueueDAO;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.util.List;
+
+@Singleton
+public class WorkflowObfuscationQueuePublisher {
```
Comment (apanicker-nflx) [inline obfuscation/src/main/java/com/netflix/conductor/obfuscation/publisher/WorkflowObfuscationQueuePublisher.java]: Should this be implementing the `WorkflowObfuscationQueuePublisher` in core module?

**Comment by andrea11** [comment]
Hello @Ismaley, may you update this PR to resolve conflicts?

**Comment by Ismaley** [comment]
@andrea11 @apanicker-nflx I'm currently working on this PR, the moment I pushed the latest changes it wasn't ready yet, there's still some things I gotta change/improve along with some changes asked. Sorry about the trouble. 
I believe I'll get this ready for another review by the end of this week or so. :)

**Comment by andrea11** [comment]
@Ismaley Hello, do you have any news on this feature?

**Comment by Ismaley** [comment]
Hi @andrea11, I've been working on this feature during this week and I believe soon I'll be able to push the changes.
Sorry that it's been taking forever, I had to switch projects inside the company and to work on these changes must be during my free time. But I'm glad to see this feature will help others and I'm excited to merge it on conductor :)

**Comment by Ismaley** [comment]
@andrea11 just updating here, I'm close to finish the changes, right now I'm trying to write an integration test case for this feature.

Plus, I tried to run conductor on my machine to test this feature and I created a simple workflow with a http task, turns out that the task remains scheduled forever and the workflow never completes. Last version of conductor I tested was 2.11, did anything changed regarding creating workflow or task definitions? Thoughts, @kishorebanala?
Tested it both with embedded and external dynomite.

**Comment by kishorebanala** [comment]
@Ismaley Not related to the way tasks are executed. Do you by any change have system tasks workers disabled? Or, do you see any errors in your logs when polling for system tasks? [SystemTaskWorkerCoordinator](https://github.com/Netflix/conductor/blob/master/core/src/main/java/com/netflix/conductor/core/execution/tasks/SystemTaskWorkerCoordinator.java) is the entry point that polls and executed system tasks.

**Comment by andrea11** [comment]
Hi @Ismaley, 
Sorry I was busy latestly. I read that you would like to improve the documentation later. But maybe it could be useful even for reviewer to better understand how this functionality works 😄 
Anyway, thanks a lot for your efforts to complete this PR. Could you try to fix your commit with the comments provided by @apanicker-nflx ?

**Comment by apanicker-nflx** [comment]
Closing this PR for now due to inactivity, please reopen when you are able to resume working on this.
